src/MapPane.js

```diff
@@ -0,0 +1,158 @@
+/* @flow */
+import React, { PropTypes, Component } from 'react'
+import { forEach, omit, uniqueId, indexOf } from 'lodash'
+
+import childrenType from './types/children'
+import mapType from './types/map'
+import paneType from './types/pane'
+
+const BLACKLIST = ['tile', 'shadow', 'overlay', 'map', 'marker', 'tooltip', 'popup']
+
+export default class MapPane extends Component {
+  static propTypes = {
+    name: PropTypes.string,
+    children: childrenType,
+    map: mapType,
+    zIndex: PropTypes.number,
+    className: PropTypes.string,
+    style: PropTypes.object,
+    pane: paneType,
+  };
+
+  static contextTypes = {
+    map: mapType,
+    pane: paneType,
+  };
+
+  static childContextTypes = {
+    pane: paneType,
+  };
+
+  constructor ({name}) {
+    super()
+
+    this.state = {
+      _isMounted: false,
+    }
+
+    this._name = name || `pane-${uniqueId()}`
+  }
+
+  getChildContext () {
+    return {
+      pane: this._name,
+    }
+  }
+
+  componentWillMount () {
+    const map = this.context.map || this.props.map
+
+    if (this._name && map && map.createPane) {
+      const existing = this.getPane()
+
+      if (!existing) {
+        map.createPane(this._name, this.getParentPane())
+      }
+    }
+  }
+
+  componentWillReceiveProps (props) {
+    this.setStyle(props)
+  }
+
+  setStyle ({ style, zIndex } = this.props) {
+    const pane = this.getPane()
+
+    if (pane) {
+      pane.style.zIndex = zIndex || 'initial'
+
+      if (style) {
+        forEach(style, (value, key) => {
+          pane.style[key] = value
+        })
+      }
+    }
+  }
+
+  componentDidMount () {
+    this.setState({
+      _isMounted: true,
+    })
+
+    this.setStyle()
+  }
+
+  componentWillUnmount () {
+    const { name } = this.props
+
+    if (name) {
+      const _name = (name + '').replace(/-*pane/gi, '')
+
+      if (indexOf(BLACKLIST, _name) >= 0) {
```

Comment:
I don't like the idea of allowing to update Leaflet's default panes unless there is a way to restore them to their original styles, that would create unexpected behavior, so I'd rather have this library simply prevent from overriding these panes.
Regarding having "`<Pane name='myCustomPane' />` in two places", same I don't think it should be possible, even if it is technically possible I think it would only lead to unexpected behavior when trying to set different styles in each component for example.